---
title: "Validation of synthetic data algorithm"
author: "Jason T. Serviss"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, message = FALSE}
packages <- c(
  "sp.scRNAseq",
  "sp.scRNAseqData",
  "sp.scRNAseqTesting",
  "seqTools",
  "printr",
  "ggthemes",
  "tidyverse"
)
purrr::walk(packages, library, character.only = TRUE)
rm(packages)
```

```{r}
#' Reorder an x or y axis within facets
#'
#' Reorder a column before plotting with faceting, such that the values are ordered
#' within each facet. This requires two functions: \code{reorder_within} applied to
#' the column, then either \code{scale_x_reordered} or \code{scale_y_reordered} added
#' to the plot.
#' This is implemented as a bit of a hack: it appends ___ and then the facet
#' at the end of each string.
#'
#' @param x Vector to reorder.
#' @param by Vector of the same length, to use for reordering.
#' @param within Vector of the same length that will later be used for faceting
#' @param fun Function to perform within each subset to determine the resulting
#' ordering. By default, mean.
#' @param sep Separator to distinguish the two. You may want to set this manually
#' if ___ can exist within one of your labels.
#' @param ... In \code{reorder_within} arguments passed on to \code{\link{reorder}}.
#' In the scale functions, extra arguments passed on to
#' \code{\link[ggplot2]{scale_x_discrete}} or \code{\link[ggplot2]{scale_y_discrete}}.
#'
#' @source "Ordering categories within ggplot2 Facets" by Tyler Rinker:
#' \url{https://trinkerrstuff.wordpress.com/2016/12/23/ordering-categories-within-ggplot2-facets/}
#'
#' @examples
#'
#' library(tidyr)
#' library(ggplot2)
#'
#' iris_gathered <- gather(iris, metric, value, -Species)
#'
#' # reordering doesn't work within each facet (see Sepal.Width):
#' ggplot(iris_gathered, aes(reorder(Species, value), value)) +
#'   geom_boxplot() +
#'   facet_wrap(~ metric)
#'
#' # reorder_within and scale_x_reordered work.
#' # (Note that you need to set scales = "free_x" in the facet)
#' ggplot(iris_gathered, aes(reorder_within(Species, value, metric), value)) +
#'   geom_boxplot() +
#'   scale_x_reordered() +
#'   facet_wrap(~ metric, scales = "free_x")
#'
#' @export
reorder_within <- function(x, by, within, fun = mean, sep = "___", ...) {
  new_x <- paste(x, within, sep = sep)
  stats::reorder(new_x, by, FUN = fun)
}

#' @rdname reorder_within
#' @export
scale_x_reordered <- function(..., sep = "___") {
  reg <- paste0(sep, ".+$")
  ggplot2::scale_x_discrete(labels = function(x) gsub(reg, "", x), ...)
}

#' @rdname reorder_within
#' @export
scale_y_reordered <- function(..., sep = "___") {
  reg <- paste0(sep, ".+$")
  ggplot2::scale_y_discrete(labels = function(x) gsub(reg, "", x), ...)
}
```

```{r}
#cpm normalization withoit adding 1 that works on both matrices and tibbles with
#genes as rows and cells as columns
norm <- function(counts) {
  if(all(class(counts) == c("tbl_df", "tbl", "data.frame"))) {
    mat <- counts %>% 
      as.data.frame() %>%
      column_to_rownames("gene") %>%
      as.matrix()
    
    t(t(mat) / colSums(mat) * 10^6) %>%
      matrix_to_tibble("gene")
  } else if(class(counts) == "matrix") {
    t(t(counts) / colSums(counts) * 10^6)
  }
}
```

The synthetic multiplets are based on the mixing of real synthetic singlets. The 
algorithm for generating the synthetic multiplets works as follows: The sum of 
counts is calculated for each gene in two randomly selected singlets from 
each cell type contributing to the multiplet. The fraction of input into the 
multiplet can be adjusted so that, e.g. cell type A only contributes half of 
what cell type B contributes to the multiplet. A vector is then generated where 
each gene name is represented one time for each of the total counts (e.g. if 
gene "A" has a sum of counts from all singlets from the cell types 
contributing to the multiplet of 10, "A" will be repreated 10 times in the 
vector). This vector is then sampled without replacment. It is sampled x times, 
where x is equal to the sum of all the counts in all singlets contributing to 
the multiplet divided by the total number of singlets contributing to the 
multiplet. Gene names are then counted which provides the raw counts for the 
newly synthesized multiplet.

Here we utilize the sorted cell line dataset to generate multiplets from the 
singlets. The synthesized multiplets are then compared to the real multiplets
to acertain the quality of the _in silico_ synthesis. We begin by synthesizing 
one multiplet per cell line combination, i.e. a A375-HCT116, A375-HOS, and 
HCT116-HOS multiplet.

### Synthesize one multiplet per combination

```{r}
s <- str_detect(colnames(countsSorted2), "^s")
sng <- countsSorted2[, s]
classes <- slice(countsSortedMeta2, match(colnames(sng), countsSortedMeta2$sample))$cellTypes
combos <- apply(combn(unique(classes), 2), 2, sort)
adjustment <- rep(1, length(unique(classes)))
names(adjustment) <- unique(classes)

synth <- sng %>%
  norm() %>%
  syntheticMultipletsFromCounts(., classes, combos, adjustment) %>%
  norm()

mul <- countsSorted2[, !s] %>%
  norm() %>%
  matrix_to_tibble("gene") %>%
  gather(sample, counts, -gene) %>%
  left_join(countsSortedMeta2, by = "sample") %>%
  select(gene, counts, cellTypes, sample)

#randomly select some genes to facilitate plotting
genes <- sample(rownames(sng), 500, replace = FALSE)
concatDat <- gather(synth, cellTypes, synCounts, -gene) %>%
  inner_join(., mul, by = c("gene", "cellTypes")) %>%
  filter(gene %in% genes)
```

Examine drop-out rate in the real data compared to the synthetic data.
```{r}
dr <- countsSorted2[, !s] %>%
  norm() %>%
  matrix_to_tibble("gene") %>%
  select(filter(countsSortedMeta2, cellTypes %in% colnames(synth))$sample) %>%
  as.matrix() %>%
  apply(., 2, function(x) length(which(x == 0))) %>%
  `/` (nrow(countsSorted2)) %>%
  mean()

print(paste0("real: ", dr))

ds <- synth %>%
  select(-gene) %>%
  as.matrix() %>%
  apply(., 2, function(x) length(which(x == 0))) %>%
  `/` (nrow(countsSorted2)) %>%
  mean()

print(paste0("synthetic: ", ds))
```

Results are shown in several different plots below. 

```{r, fig.align="center", fig.width=8, fig.height=6}
#linear plot
concatDat %>%
  ggplot() +
  geom_point(aes(reorder_within(gene, counts, cellTypes, fun = "mean"), counts), size = 0.1) +
  geom_line(
    data = . %>% select(gene, cellTypes, synCounts) %>% distinct(),
    aes(reorder_within(gene, synCounts, cellTypes, fun = "mean"), synCounts, group = cellTypes),
    colour = "red"
  ) +
  facet_wrap(~cellTypes) +
  scale_x_reordered() +
  theme_few() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "500 randomly selected genes", y = "cpm")
```

```{r, fig.align="center", fig.width=8, fig.height=6}
#log plot
concatDat %>%
  ggplot() +
  geom_point(aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1) +
  geom_point(
    data = . %>% select(gene, cellTypes, synCounts) %>% distinct(),
    aes(reorder_within(gene, log2(synCounts + 1), cellTypes, fun = "mean"), log2(synCounts + 1)),
    colour = "red", size = 0.3
  ) +
  facet_wrap(~cellTypes, scales = "free") +
  scale_x_reordered() +
  theme_few() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "500 randomly selected genes", y = "log2(cpm)")
```

```{r, fig.align="center", fig.width=8, fig.height=6}
#violin plot
concatDat %>%
  filter(gene %in% sample(concatDat$gene, 50, replace = FALSE)) %>%
  ggplot() +
  geom_violin(aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1) +
  geom_point(aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1) +
  geom_point(
    data = . %>% select(gene, cellTypes, synCounts) %>% distinct(),
    aes(reorder_within(gene, log2(synCounts + 1), cellTypes, fun = "mean"), log2(synCounts + 1)),
    colour = "red", size = 0.3
  ) +
  facet_wrap(~cellTypes, scales = "free") +
  scale_x_reordered() +
  theme_few() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "50 randomly selected genes", y = "log2(cpm)")
```

<br></br>

### Synthesize 10 multiplets per combination.

Here we continue to examine the ability of the synthesized multiplets to 
accuratley reflect the real multiplets by generating 10 multiplets per 
combination. It is hoped that this will show that the synthesized multiplets as
a whole, accuratley capture the diversity of the real multiplets.
```{r}
n <- 10
names <- paste(
  apply(combos, 2, paste, collapse = "-"), 
  rep(1:n, each = ncol(combos)), 
  sep = "."
)

synth <- sng %>%
  norm() %>%
  {map(1:n, function(x) {
    syntheticMultipletsFromCounts(., classes, combos, adjustment, seed = x + 9789)
  })} %>%
  reduce(full_join, by = "gene") %>%
  setNames(c("gene", names)) %>%
  norm()

genes <- sample(rownames(sng), 500, replace = FALSE)
concatDat <- synth %>% 
  filter(gene %in% genes) %>%
  gather(typeRep, synCounts, -gene) %>%
  separate(typeRep, c("cellTypes", "sample"), sep = "\\.") %>%
  inner_join(., mul, by = c("gene", "cellTypes"))
```

Examine drop-out rate.
```{r}
dr <- countsSorted2[, !s] %>%
  norm() %>%
  matrix_to_tibble("gene") %>%
  select(filter(countsSortedMeta2, cellTypes %in% concatDat$cellTypes)$sample) %>%
  as.matrix() %>%
  apply(., 2, function(x) length(which(x == 0))) %>%
  `/` (nrow(countsSorted2)) %>%
  mean()

print(paste0("real: ", dr))

ds <- synth %>%
  select(-gene) %>%
  as.matrix() %>%
  apply(., 2, function(x) length(which(x == 0))) %>%
  `/` (nrow(countsSorted2)) %>%
  mean()

print(paste0("synthetic: ", ds))
```

Plot results.

```{r, fig.align="center", fig.width=8, fig.height=6}
#linear
concatDat %>%
  ggplot() +
  geom_point(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, counts, cellTypes, fun = "mean"), counts), size = 0.1
  ) +
  geom_line(
    data = . %>% select(gene, cellTypes, synCounts, sample.x) %>% distinct(),
    aes(reorder_within(gene, synCounts, cellTypes, fun = "mean"), synCounts, group = cellTypes, colour = sample.x)
  ) +
  facet_wrap(~cellTypes, scales = "free") +
  theme_few() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "500 randomly selected genes", y = "cpm") +
  guides(colour = FALSE)
```

```{r, fig.align="center", fig.width=8, fig.height=6}
#log
concatDat %>%
  ggplot() +
  geom_point(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, synCounts, sample.x) %>% distinct(),
    aes(reorder_within(gene, log2(synCounts + 1), cellTypes, fun = "mean"), log2(synCounts + 1), colour = sample.x), 
    size = 0.3
  ) +
  facet_wrap(~cellTypes, scales = "free") +
  theme_few() +
  theme(
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank()
  ) +
  labs(x = "500 randomly selected genes", y = "log2(cpm)") +
  guides(colour = FALSE)
```


```{r, fig.align="center", fig.width=8, fig.height=6}
#violin
concatDat <- synth %>% 
  filter(gene %in% sample(concatDat$gene, 50, replace = FALSE)) %>%
  gather(typeRep, synCounts, -gene) %>%
  separate(typeRep, c("cellTypes", "sample"), sep = "\\.") %>%
  inner_join(., mul, by = c("gene", "cellTypes"))

concatDat %>%
  filter(cellTypes == "HCT116-HOS") %>%
  ggplot() +
  geom_violin(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, synCounts, sample.x) %>% distinct(),
    aes(reorder_within(gene, log2(synCounts + 1), cellTypes, fun = "mean"), log2(synCounts + 1), colour = sample.x), 
    size = 0.3
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size = 3)) +
  labs(x = "50 randomly selected genes", y = "log2(cpm)", caption = "HCT116-HOS") +
  guides(colour = FALSE)

concatDat %>%
  filter(cellTypes == "A375-HCT116") %>%
  ggplot() +
  geom_violin(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, synCounts, sample.x) %>% distinct(),
    aes(reorder_within(gene, log2(synCounts + 1), cellTypes, fun = "mean"), log2(synCounts + 1), colour = sample.x), 
    size = 0.3
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size = 3)) +
  labs(x = "50 randomly selected genes", y = "log2(cpm)", caption = "A375-HCT116") +
  guides(colour = FALSE)

concatDat %>%
  filter(cellTypes == "A375-HOS") %>%
  ggplot() +
  geom_violin(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, counts, sample.y) %>% distinct(),
    aes(reorder_within(gene, log2(counts + 1), cellTypes, fun = "mean"), log2(counts + 1)), size = 0.1
  ) +
  geom_point(
    data = . %>% select(gene, cellTypes, synCounts, sample.x) %>% distinct(),
    aes(reorder_within(gene, log2(synCounts + 1), cellTypes, fun = "mean"), log2(synCounts + 1), colour = sample.x), 
    size = 0.3
  ) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 90, size = 3)) +
  labs(x = "50 randomly selected genes", y = "log2(cpm)", caption = "A375-HOS") +
  guides(colour = FALSE)
```

Run tsne with real and synthesized multiplets.
```{r, fig.align="center", fig.width=8, fig.height=6}
#tsne
set.seed(23435)
countsSorted2[, colnames(countsSorted2) %in% filter(countsSortedMeta2, cellTypes %in% concatDat$cellTypes)$sample] %>%
  norm() %>% 
  matrix_to_tibble("gene") %>%
  full_join(synth, by = "gene") %>%
  as.data.frame() %>%
  column_to_rownames("gene") %>%
  as.matrix() %>%
  `+` (1) %>%
  log2(.) %>%
  pearsonsCor(., select = nTopMax(., 2000)) %>%
  Rtsne::Rtsne(., is_distance = TRUE, perplexity = 15) %>%
  `[[` ('Y') %>% 
  matrix_to_tibble() %>%
  add_column(sample = c(colnames(synth[, -1]), colnames(countsSorted2)[colnames(countsSorted2) %in% filter(countsSortedMeta2, cellTypes %in% concatDat$cellTypes)$sample])) %>%
  mutate(type = if_else(str_detect(sample, "^m"), "real", "synthetic")) %>%
  ggplot() +
  geom_point(aes(V1, V2, colour = type))
```

<br></br>

### Adjusting multiplet contribution

Synthesize 25 multiplets for each combination and reduce the contribution
of HOS cells by 0%, 25%, 50%, 75%, or 99%. Calculate the mean gene expression 
for each and plot together with the real multiplets and corresponding singlets.

```{r, fig.align="center", fig.width=10, fig.height=8, message = FALSE}
###FUNCTIONS
#function to calculate the mean gene expression for a specific cell type. Used 
#previous to heatmap plotting
typeMeans <- function(type) {
  countsSortedMeta2 %>%
  filter(cellTypes == type) %>% 
  pull(sample) %>%
  {countsSorted2[, colnames(countsSorted2) %in% .]} %>%
  norm() %>%
  rowMeans() %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  setNames(c("gene", type)) %>%
  as_tibble()
}

#wrapper to generate and reformat multiplets
generateAdjustedMultiplets <- function(sng, classes, combos, adjustment, name) {
  q.name <- enquo(name)
  sng %>%
    norm() %>%
    {map(1:n, function(x) {
      syntheticMultipletsFromCounts(., classes, combos, adjustment, seed = x + 9789)
    })} %>%
    reduce(full_join, by = "gene") %>%
    setNames(c("gene", names)) %>%
    norm() %>%
    gather(cellTypes, !!q.name, -gene) %>%
    mutate(cellTypes = str_replace(cellTypes, "(.*)\\..*", "\\1")) %>%
    group_by(gene, cellTypes) %>%
    summarize(!! quo_name(q.name) := mean(!!q.name)) %>%
    ungroup()
}
###

#Synthesize multiplets with varying levels of adjusted HOS input
adjustment <- rep(1, length(unique(classes)))
names(adjustment) <- unique(classes)
n <- 25
names <- paste(
  apply(combos, 2, paste, collapse = "-"), 
  rep(1:n, each = ncol(combos)), 
  sep = "."
)

synth <- generateAdjustedMultiplets(sng, classes, combos, adjustment, synCounts)
adjustment[names(adjustment) == "HOS"] <- 0.75
synth7 <- generateAdjustedMultiplets(sng, classes, combos, adjustment, synth7)
adjustment[names(adjustment) == "HOS"] <- 0.5
synth5 <- generateAdjustedMultiplets(sng, classes, combos, adjustment, synth5)
adjustment[names(adjustment) == "HOS"] <- 0.25
synth25 <- generateAdjustedMultiplets(sng, classes, combos, adjustment, synth25)
adjustment[names(adjustment) == "HOS"] <- 0.01
synth01 <- generateAdjustedMultiplets(sng, classes, combos, adjustment, synth01)

#heatmap
#find overexpressed genes in each cell type
genes <- foldChangePerClass(cpm(sng), rename(countsSortedMeta2, class = cellTypes)) %>%
  matrix_to_tibble("gene") %>%
  select(gene, A375, HCT116, HOS) %>%
  gather(cellType, fold, -gene) %>%
  group_by(cellType) %>%
  top_n(25) %>%
  ungroup()

#calculate the mean gene expression for each type of singlet
a375 <- typeMeans("A375") %>% mutate(cellType = "A375") %>% rename(counts = A375)
hct116 <- typeMeans("HCT116") %>% mutate(cellType = "HCT116") %>% rename(counts = HCT116)
hos <- typeMeans("HOS") %>% mutate(cellType = "HOS") %>% rename(counts = HOS)
realSng <- bind_rows(a375, hct116, hos)

#calculate the mean gene expression for each type of multiplet
a375_hct116 <- typeMeans("A375-HCT116") %>% 
  add_column(cellTypes = "A375-HCT116") %>%
  rename(counts = `A375-HCT116`)
a375_hos <- typeMeans("A375-HOS") %>% 
  add_column(cellTypes = "A375-HOS") %>%
  rename(counts = `A375-HOS`)
hct116_hos <- typeMeans("HCT116-HOS") %>% 
  add_column(cellTypes = "HCT116-HOS") %>%
  rename(counts = `HCT116-HOS`)

#concatenate real singlet, real multiplet, and info for genes corresponding to 
#a cell type
real <- bind_rows(a375_hct116, a375_hos, hct116_hos) %>%
  rename(multipletCounts = counts) %>% 
  mutate(
    cellType1 = str_replace(cellTypes, "(.*)-.*", "\\1"), 
    cellType2 = str_replace(cellTypes, ".*-(.*)", "\\1")
  ) %>% 
  inner_join(realSng, by = c("cellType1" = "cellType", "gene" = "gene")) %>% 
  inner_join(realSng, by = c("cellType2" = "cellType", "gene" = "gene")) %>%
  inner_join(genes, by = "gene") %>%
  filter(cellType == cellType1 | cellType == cellType2) %>%
  rename(geneType = cellType)

#join everything and plot
real %>%
  inner_join(synth, by = c("gene", "cellTypes")) %>%
  inner_join(synth7, by = c("gene", "cellTypes")) %>%
  inner_join(synth5, by = c("gene", "cellTypes")) %>%
  inner_join(synth25, by = c("gene", "cellTypes")) %>%
  inner_join(synth01, by = c("gene", "cellTypes")) %>%
  gather(type, cpm, -gene, -(geneType:fold), -(cellTypes:cellType2)) %>%
  mutate(type = case_when(
    type == "counts.x" ~ cellType1, 
    type == "counts.y" ~ cellType2, 
    TRUE ~ type
  )) %>%
  mutate(type = case_when(
    type == "multipletCounts" ~ "Multiplet",
    type == "synCounts" ~ "0%",
    type == "synth7" ~ "25%",
    type == "synth5" ~ "50%",
    type == "synth25" ~ "75%",
    type == "synth01" ~ "99%",
    TRUE ~ type
  )) %>%
  mutate(type = parse_factor(
    type, 
    levels = c("Multiplet", "0%", "25%", "50%", "75%", "99%", "A375", "HCT116", "HOS")
  )) %>%
  arrange(geneType, fold) %>%
  mutate(plotGene = paste0(gene, " (", geneType, ")")) %>%
  mutate(plotGene = parse_factor(plotGene, levels = unique(plotGene))) %>%
  mutate(cellTypes = parse_factor(cellTypes, levels = c("A375-HOS", "HCT116-HOS", "A375-HCT116"))) %>%
  ggplot() +
  geom_tile(aes(type, plotGene, fill = log2(cpm + 1))) +
  facet_wrap(~cellTypes, scale = "free") +
  viridis::scale_fill_viridis() +
  theme_few() +
  theme(
    axis.text.x = element_text(angle = 90),
    axis.title.x = element_blank(),
    legend.position = "top"
  ) +
  labs(y = "Gene (Cell type)")
```

<br></br>

### Synthesize multiplets with mouse data

Synthesize colon and small intestine stem multiplets with the mouse data.
```{r, fig.align="center", fig.width=8, fig.height=6}
###FUNCTIONS
plotTSNEgene <- function(geneToPlot, slog, cObjMul) {
  s <- slog %>%
    filter(gene == geneToPlot) %>% 
    select(-gene) %>% 
    rename(gene = counts)
  
  getData(cObjMul, "counts.log")[geneToPlot, ] %>%
    as.data.frame() %>%
    rownames_to_column() %>%
    setNames(c("sample", "gene")) %>%
    as_tibble() %>%
    bind_rows(s) %>%
    full_join(tresults, by = "sample") %>%
    mutate(type = if_else(str_detect(sample, "^m"), "real", "synthetic")) %>%
    ggplot() +
    geom_point(aes(V1, V2, colour = gene, shape = type)) +
    viridis::scale_color_viridis() +
    guides(colour = guide_colorbar(title = geneToPlot))
}
###

s <- str_detect(colnames(countsMgfp), "^s")
commonGenes <- intersect(rownames(countsMgfp), rownames(countsRegev))

sng <- cbind(countsMgfp[commonGenes, s], countsRegev[commonGenes, ])
mul <- countsMgfp[commonGenes, !s]

erccSng <- cbind(
  countsMgfpERCC[, s], 
  matrix(NA, nrow = nrow(countsMgfpERCC), ncol = ncol(countsRegev))
)
erccMul <- cbind(countsMgfpERCC[, !s])

boolMulC <- colnames(mul) %in% filter(countsMgfpMeta, tissue == "colon")$sample
boolSngC <- colnames(sng) %in% filter(countsMgfpMeta, tissue == "colon")$sample

#setup spCounts
cObjSng <- spCounts(sng, erccSng)
cObjMul <- spCounts(mul, erccMul)

#spUnsupervised
uObj <- spUnsupervised(cObjSng, seed = 2334, max_iter = 1000, max = 2000)
plotUnsupervisedClass(uObj, cObjSng) %>%
  plotData() %>%
  ggplot() +
  geom_point(
    aes(
      `t-SNE dim 1`, `t-SNE dim 2`, 
      colour = Classification, size = Uncertainty
    )
  ) + 
  geom_label(
    data = getData(uObj, "tsneMeans"),
    aes(x = x, y = y, label = classification)
  )

classes <- getData(uObj, "classification")

bool <- classes %in% c("B1", "K1", "R1", "T1", "F2")
c <- case_when(
  classes %in% c("B1", "K1") ~ "Colon.Stem",
  classes %in% c("R1", "T1", "F2") ~ "SI.Stem",
  TRUE ~ classes
)
combos <- apply(combn(unique(c[bool]), 2), 2, sort)
adjustment <- rep(1, length(unique(c[bool])))
names(adjustment) <- unique(c[bool])

n <- 30
names <- paste(
  apply(combos, 2, paste, collapse = "-"), 
  rep(1:n, each = ncol(combos)), 
  sep = "."
)

synth <- getData(cObjSng, "counts")[, bool] %>%
  norm() %>%
  {map(1:n, function(x) {
    syntheticMultipletsFromCounts(., c[bool], combos, adjustment, seed = x + 9789)
  })} %>%
  reduce(full_join, by = "gene") %>%
  setNames(c("gene", names)) %>%
  norm()

#tsne
set.seed(2343)
tresults <- getData(cObjMul, "counts") %>%
  norm() %>%
  matrix_to_tibble("gene") %>%
  full_join(synth, by = "gene") %>%
  as.data.frame() %>%
  column_to_rownames("gene") %>%
  as.matrix() %>%
  `+` (1) %>%
  log2(.) %>%
  pearsonsCor(., select = nTopMax(., 1000)) %>%
  Rtsne::Rtsne(., is_distance = TRUE, max_iter = 3000) %>%
  `[[` ('Y') %>% 
  matrix_to_tibble() %>%
  select(-rowname) %>%
  add_column(sample = c(colnames(getData(cObjMul, "counts")), colnames(synth[, -1])))

#plot results
tresults %>%
  mutate(type = if_else(str_detect(sample, "^m"), "real", "synthetic")) %>%
  ggplot() +
  geom_point(aes(V1, V2, colour = type))

slog <- synth %>% 
  as.data.frame() %>%
  column_to_rownames("gene") %>%
  as.matrix() %>%
  `+` (1) %>%
  log2() %>%
  matrix_to_tibble("gene") %>%
  gather(sample, counts, -gene)

plotTSNEgene("Lgr5", slog, cObjMul)
plotTSNEgene("Lyz1", slog, cObjMul)
plotTSNEgene("Alpi", slog, cObjMul)
plotTSNEgene("Muc2", slog, cObjMul)
plotTSNEgene("Slc40a1", slog, cObjMul)
```

```{r}
sessionInfo()
```
