---
title: "Testing new swarm algo"
author: "Jason T. Serviss"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, message = FALSE}
packages <- c(
    "sp.scRNAseq",
    "sp.scRNAseqData",
    "printr",
    "ggthemes",
    "tidyverse"
)
purrr::walk(packages, library, character.only = TRUE)
rm(packages)
```

```{r}
syntheticMultipletsFromCounts <- function(
  singlets, classes, fractions,
  seed, ...
){
  set.seed(seed)
  
  #select one cell from the pool for each cell type
  idx <- purrr::map_int(unique(classes), ~sample(which(classes == .x), 1))
  exCounts <- singlets[, idx]
  
  #adjust the counts by multiplying with the corresponding values adjustment
  adjusted <- round(t(t(exCounts) * fractions))
  
  #calculate the sum of counts for each gene
  rs <- matrixStats::rowSums2(adjusted)
  
  #Sample from the poisson distribution for each gene with lambda = rs
  matrix(
    rpois(n = length(rs), lambda = rs),
    dimnames = list(rownames(exCounts), "multiplet")
  )
}

generateSyntheticMultiplets <- function(
  singlets, classes, fractions,
  seed, n, ...
){
  .norm.counts <- function(counts) {
    t(t(counts) / matrixStats::colSums2(counts) * 10^6 + 1)
  }
  
  mat <- purrr::map(1:n, function(y) {
    syntheticMultipletsFromCounts(
      singlets = singlets, classes = classes,
      fractions = fractions, seed = seed + y
    )
  }) %>%
  do.call(cbind, .) %>%
  .norm.counts()
  
  colnames(mat) <- paste0("multiplet.", 1:n)
  mat
}
```

### Use sorted cell line dataset.
Data
```{r, fig.align="center", fig.width=10, fig.height=8}
##DATA
s <- grepl("^s", colnames(countsSorted2))
cObjSng <- spCounts(countsSorted2[,s], countsSortedERCC2[,s])
cObjMul <- spCounts(countsSorted2[,!s], countsSortedERCC2[,!s])
uObj <- spUnsupervised(cObjSng)

#rename classes
positions <- str_extract(colnames(getData(cObjSng, "counts")), "...$")
newClass <- case_when(
  positions == "E03" ~ "A375", #what is this? sorting issue?
  positions %in% paste0(sort(rep(LETTERS[1:8], 4)), c("01", "02", "03", "04")) ~ "HOS",
  positions %in% paste0(sort(rep(LETTERS[1:8], 4)), c("05", "06", "07", "08")) ~ "HCT116",
  positions %in% paste0(sort(rep(LETTERS[1:8], 4)), c("09", "10", "11", "12")) ~ "A375",
  TRUE ~ "error"
)
corresp <- getData(uObj, "classification") %>%
  tibble(oldClass = ., newClass = newClass) %>%
  distinct()

classification(uObj) <- newClass

gm <- getData(uObj, "groupMeans")
colnames(gm) <- pull(corresp, newClass)[match(colnames(gm), pull(corresp, oldClass))]
groupMeans(uObj) <- gm

tm <- getData(uObj, "tsneMeans")
tm$classification <- pull(corresp, newClass)[match(tm$classification, pull(corresp, oldClass))]
tsneMeans(uObj) <- tm

plotUnsupervisedClass(uObj, cObjSng)
```

Make synthetic multiplets.
```{r}
#generate synthetic multiplets; 10 per combination
synthetic.mul <- generateSyntheticMultiplets(
  getData(cObjSng, "counts.cpm"), getData(uObj, "classification"), 
  fractions = c(0, 0.5, 0.5), seed = 8923, n = 100
)
```

Plot 
```{r, fig.align="center", fig.width=10, fig.height=8}
g <- c("ANXA3", "CD74")
synthetic.mul %>%
  matrix_to_tibble("gene") %>%
  filter(gene %in% g) %>%
  gather(sample, value, -gene) %>%
  mutate(q = list(0:2000)) %>%
  mutate(dpois = map2(q, value, ~dpois(.x, .y))) %>%
  unnest() %>%
  ggplot() +
  geom_histogram(aes(value, y = ..density..), binwidth = 1) +
  geom_line(
    aes(q, dpois, group = sample), colour = "grey", size = 0.1
  ) +
  stat_density(
    data = getData(cObjSng, "counts.cpm")[g, class == "A375"] %>% 
      matrix_to_tibble("gene") %>%
      gather(sample, value, -gene),
    aes(value, y = ..density..), geom = "line", colour = "blue"
  ) +
  stat_density(
    data = getData(cObjSng, "counts.cpm")[g, class == "HCT116"] %>% 
      matrix_to_tibble("gene") %>%
      gather(sample, value, -gene),
    aes(value, y = ..density..), geom = "line", colour = "yellow"
  ) +
  geom_line(
    data = . %>% group_by(gene, q) %>% summarize(mean = mean(dpois)),
    aes(q, mean), colour = "purple"
  ) +
  facet_wrap(~gene) +
  theme_bw() +
  labs(y = "Probability", x = "Gene expression value")
```

```{r, eval = FALSE}
#show one multiplet
#m.NJB00204.A02 is a A375-HCT116 multiplet
#genes: HOS = COL3A1, HCT116 = ANXA3, A375 = CD74

g <- c("ANXA3", "CD74", "COL3A1")
dp1 <- getData(cObjMul, "counts.cpm")[, 'm.NJB00204.A02'] %>%
  round() %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  as_tibble() %>%
  setNames(c("gene", "oneMultiplet")) %>%
  full_join(
    generateSyntheticMultiplets(
      getData(cObjSng, "counts.cpm"), getData(uObj, "classification"),
      fractions = c(0, 0.5, 0.5), seed = 8923, n = 100
    ) %>%
      matrix_to_tibble("gene"),
    by = "gene"
  ) %>%
  gather(syntheticMultipletName, syntheticMultipletValue, -gene, -oneMultiplet) %>%
  mutate(dpMultiplet = map2_dbl(oneMultiplet, syntheticMultipletValue, function(x, y) dpois(x, y))) %>%
  filter(gene %in% g) %>%
  add_column(fraction = "Correct fraction")

dp2 <- getData(cObjMul, "counts.cpm")[, 'm.NJB00204.A02'] %>%
  round() %>%
  as.data.frame() %>%
  rownames_to_column("gene") %>%
  as_tibble() %>%
  setNames(c("gene", "oneMultiplet")) %>%
  full_join(
    generateSyntheticMultiplets(
      getData(cObjSng, "counts.cpm"), getData(uObj, "classification"),
      fractions = c(1/3, 1/3, 1/3), seed = 8923, n = 100
    ) %>%
      matrix_to_tibble("gene"),
    by = "gene"
  ) %>%
  gather(syntheticMultipletName, syntheticMultipletValue, -gene, -oneMultiplet) %>%
  mutate(dpMultiplet = map2_dbl(oneMultiplet, syntheticMultipletValue, ~dpois(.x, .y))) %>%
  filter(gene %in% g) %>%
  add_column(fraction = "Incorrect fraction")

expression <- getData(cObjSng, "counts.cpm")[g, ] %>% 
  matrix_to_tibble("gene") %>%
  gather(sample, value, -gene) %>%
  add_column(class = rep(getData(uObj, "classification"), each = length(g))) %>%
  mutate(gene = case_when(
    gene == "ANXA3" ~ "ANXA3 (HCT116) ",
    gene == "COL3A1" ~ "COL3A1 (HOS)",
    gene == "CD74" ~ "CD74 (A375)",
    TRUE ~ "error"
  ))

bind_rows(dp1, dp2) %>% 
  mutate(gene = case_when(
    gene == "ANXA3" ~ "ANXA3 (HCT116) ",
    gene == "COL3A1" ~ "COL3A1 (HOS)",
    gene == "CD74" ~ "CD74 (A375)",
    TRUE ~ "error"
  )) %>%
  mutate(
    q = list(0:1100),
    dpSynthetic = map2(
    q, syntheticMultipletValue, 
    ~dpois(.x, .y)
  )) %>%
  unnest() %>%
  ggplot() +
  geom_histogram(
    data = expression,
    aes(value, y = ..density.., fill = class), alpha = 1, 
    position = position_dodge(), binwidth = 100
  ) +
  geom_line(
    aes(q, dpSynthetic, group = syntheticMultipletName), 
    colour = "grey", alpha = 0.4, size = 0.1
  ) +
  geom_point(
    data = . %>% 
      select(gene, syntheticMultipletValue, dpMultiplet, fraction) %>% 
      distinct(),
    aes(syntheticMultipletValue, dpMultiplet, group = fraction), 
    size = 0.1, colour = "red"
  ) +
  geom_text(
    data = . %>% group_by(fraction, gene) %>% summarize(mean = -log10(mean(dpMultiplet))),
    aes(x = 10, y = 0.4, label = paste0("cost: ", round(mean, digits = 2))),
    hjust = 0
  ) +
  facet_grid(fraction~gene, scales = "free_y") +
  ggthemes::theme_few() +
  ggthemes::scale_fill_ptol() +
  theme(legend.position = "top") +
  labs(x = "Gene expression value", y = "Density") +
  guides(fill = guide_legend(title = "Cell type"))
```

### Use mouse small intestine dataset.
Data
```{r, fig.align="center", fig.width=10, fig.height=8}
s <- str_detect(colnames(countsMgfp), "^s")
commonGenes <- intersect(rownames(countsMgfp), rownames(countsRegev))

sng <- cbind(countsMgfp[commonGenes, s], countsRegev[commonGenes, ])
mul <- countsMgfp[commonGenes, !s]

erccSng <- cbind(
  countsMgfpERCC[, s], 
  matrix(NA, nrow = nrow(countsMgfpERCC), ncol = ncol(countsRegev))
)
erccMul <- cbind(countsMgfpERCC[, !s])

boolMulC <- colnames(mul) %in% filter(countsMgfpMeta, tissue == "colon")$sample
boolSngC <- colnames(sng) %in% filter(countsMgfpMeta, tissue == "colon")$sample

#setup spCounts
cObjSngSi <- spCounts(sng[, !boolSngC], erccSng[, !boolSngC])
cObjMulSi <- spCounts(mul[, !boolMulC], erccMul[, !boolMulC])

#spUnsupervised
uObjSi <- spUnsupervised(cObjSngSi, max_iter = 7000, initial_dims = sum(!boolSngC), max = 1000, seed = 89142)
#plotUnsupervisedClass(uObjSi, cObjSngSi)

#rename
siClass <- tibble(
  sample = rownames(getData(uObjSi, "tsne")),
  oldClass = getData(uObjSi, "classification")
) %>%
  mutate(newClass = case_when(
    oldClass %in% c("N1", "E1", "M1", "P1", "Q1", "G1", "D1", "I1", "F1", "H1") ~ "SI.Stem",
    oldClass %in% c("C1", "R1", "J1") ~ "SI.Enterocytes",
    oldClass == "K1" ~ "SI.Endocrine",
    oldClass == "A1" ~ "Blood",
    oldClass == "L1" ~ "SI.Tuft.Cell",
    oldClass %in% c("B1", "O1") ~ "SI.Goblet",
    oldClass == "S1" ~ "SI.Paneth",
    TRUE ~ "error"
  ))

classification(uObjSi) <- siClass$newClass[match(rownames(getData(uObjSi, "tsne")), siClass$sample)]
groupMeans(uObjSi) <- averageGroupExpression(cObjSngSi, getData(uObjSi, "classification"), FALSE)
tsneMeans(uObjSi) <- tsneGroupMeans(getData(uObjSi, "tsne"), getData(uObjSi, "classification"))
plotUnsupervisedClass(uObjSi, cObjSngSi)
```

Make synthetic multiplets
```{r}
#make multiplets
cpm <- getData(cObjSngSi, "counts.cpm")
class <- getData(uObjSi, "classification")
adjust <- c(0.5, 0, 0, 0.5, 0, 0, 0)
names(adjust) <- unique(getData(uObjSi, "classification"))
n.multiplets <- 100

#generate synthetic multiplets; 10 per combination
synthetic.mul <- map(1:n.multiplets, function(y) {
  syntheticMultipletsFromCounts(cpm, class, adjust, seed = 3948 + y)
}) %>% 
  purrr::reduce(cbind) %>%
  seqTools::cpm()
```

Plot
```{r, fig.align="center", fig.width=10, fig.height=8}
g <- c("Lgr5", "Ptprc", "Actb", "Dclk1")
synthetic.mul %>%
  matrix_to_tibble("gene") %>%
  filter(gene %in% g) %>%
  gather(sample, value, -gene) %>%
  group_by(gene) %>% 
  mutate(q = list(0:max(value))) %>%
  ungroup() %>%
  mutate(dpois = map2(q, value, ~dpois(.x, .y))) %>%
  unnest() %>%
  ggplot() +
  geom_histogram(data = . %>% filter(gene == "Lgr5"), aes(value, y = ..density..), binwidth = 20) +
  geom_histogram(data = . %>% filter(gene == "Ptprc"), aes(value, y = ..density..), binwidth = 50) +
  geom_histogram(data = . %>% filter(gene == "Actb"), aes(value, y = ..density..), binwidth = 50) +
  geom_histogram(data = . %>% filter(gene == "Dclk1"), aes(value, y = ..density..), binwidth = 0.1) +
  geom_line(
    aes(q, dpois, group = sample), colour = "grey", size = 0.1, alpha = 1/2
  ) +
  geom_line(
    data = . %>% group_by(gene, q) %>% summarize(mean = mean(dpois)),
    aes(q, mean, group = gene), colour = "purple", size = 0.3
  ) +
  facet_wrap(~gene, scales = "free") +
  ggthemes::theme_few() +
  labs(y = "Probability", x = "Gene expression value")
```

```{r}
sessionInfo()
```


