---
title: "Mouse colon analysis, 20 classes"
author: "Jason T. Serviss"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: yes
    code_folding: hide
editor_options: 
  chunk_output_type: console
---

```{r, message = FALSE}
packages <- c("CIMseq", "CIMseq.testing", "tidyverse", "circlize", "printr")
purrr::walk(packages, library, character.only = TRUE)
rm(packages)

##DATA
load('../data/CIMseqData.rda')
load('../data/sObj.rda')
load('../data/seuratDE.rda')
if(!dir.exists('../figures')) dir.create('../figures')

#there are 5 cells that were classified as colon but sorted as SI. These have to
#be removed manually
c <- getData(cObjSng, "classification")
s <- names(c[c %in% c("4", "9")])
i <- which(colnames(getData(cObjSng, "counts")) %in% s)
cObjSng <- CIMseqSinglets(
  getData(cObjSng, "counts")[, -i],
  getData(cObjSng, "counts.ercc")[, -i],
  getData(cObjSng, "dim.red")[-i, ],
  getData(cObjSng, "classification")[-i]
)
```

```{r, eval = TRUE, echo = FALSE}
#rename classes
getData(cObjSng, "classification") <- renameClasses.MGA(getData(cObjSng, "classification"), FALSE)
fractions <- getData(sObj, "fractions")
fractions <- fractions[, !colnames(fractions) %in% c("4", "9")]
colnames(fractions) <- renameClasses.MGA(colnames(fractions), FALSE)
sObj@fractions <- fractions
```

### Fig 1: Classes
```{r, fig.align="center", fig.width=10, fig.height=8}
p <- plotUnsupervisedClass(cObjSng, cObjMul, palette('c'))
p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.classes.pdf',
  device = cairo_pdf,
  height = 180,
  width = 180,
  units = "mm"
)

```

### Fig 2: Cell type gene expression
```{r, fig.align="center", fig.width=10, fig.height=8}
p <- plotUnsupervisedMarkers(
  cObjSng, cObjMul,
  c("Lgr5", "Ptprc", "Chga", "Dclk1", "Slc26a3", "Atoh1"),
  pal = RColorBrewer::brewer.pal(8, "Set1")
)
p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.markers.pdf',
  device = cairo_pdf,
  height = 180,
  width = 180,
  units = "mm"
)

```

### Fig 3: Cell cycle and Plet1
```{r, fig.align="center", fig.width=10, fig.height=8}
p <- plotUnsupervisedMarkers(
  cObjSng, cObjMul, c("Mki67"),
  pal = RColorBrewer::brewer.pal(8, "Set1")
)
p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.Mki67.pdf',
  device = cairo_pdf,
  height = 180,
  width = 180,
  units = "mm"
)
#Shown in differential expression analysis instead
# p <- plotUnsupervisedMarkers(
#   cObjSng, cObjMul, c("Plet1"),
#   pal = RColorBrewer::brewer.pal(8, "Set1")
# )
# p
# ggsave(
#   plot = p,
#   filename = '../figures/MGA.enge20.Plet1.pdf',
#   device = cairo_pdf,
#   height = 180,
#   width = 180,
#   units = "mm"
# )
```

### Fig 4: Connections per multiplet
```{r}
adj <- adjustFractions(cObjSng, cObjMul, sObj)
as.data.frame(table(apply(adj, 1, sum)))
```

### Fig 5: Fraction histogram
```{r, fig.align="center", fig.width=10, fig.height=8}
tibble(fractions = c(fractions)) %>%
  ggplot() +
  geom_histogram(aes(fractions), binwidth = 0.01) +
  theme_bw()
```

### Fig 6: Detected cell types vs. cost
```{r, fig.align="center", fig.width=10, fig.height=8}
tibble(
  nCellTypes = apply(adj, 1, sum),
  cost = getData(sObj, "costs")
) %>%
  ggplot() +
  geom_boxplot(aes(nCellTypes, cost, group = nCellTypes)) +
  scale_x_continuous(name = "Detected cell types", breaks = 0:max(apply(adj, 1, sum))) +
  theme_bw()
```

### Fig 7: Estimated cell numbers vs. cost
```{r, fig.align="center", fig.width=10, fig.height=8}
tibble(
  sample = names(getData(sObj, "costs")),
  cost = unname(getData(sObj, "costs"))
) %>%
  inner_join(
    select(estimateCells(cObjSng, cObjMul), sample, estimatedCellNumber), 
    by = "sample"
  ) %>%
  mutate(estimatedCellNumber = round(estimatedCellNumber)) %>%
  ggplot() +
  geom_boxplot(aes(estimatedCellNumber, cost, group = estimatedCellNumber)) +
  scale_x_continuous(
    name = "ERCC estimated cell number", 
    breaks = 0:max(round(pull(estimateCells(cObjSng, cObjMul), estimatedCellNumber)))
  ) +
  theme_bw()
```

### Fig 8: Estimated cell number vs. Detected cell number
```{r, fig.align="center", fig.width=10, fig.height=8}
ercc <- filter(estimateCells(cObjSng, cObjMul), sampleType == "Multiplet")
nConnections <- apply(adj, 1, sum)
nConnections <- nConnections[match(ercc$sample, names(nConnections))]
tibble(
  detectedConnections = round(nConnections),
  estimatedCellNumber = round(ercc$estimatedCellNumber)
) %>%
  ggplot() +
  geom_boxplot(aes(estimatedCellNumber, detectedConnections, group = estimatedCellNumber)) +
  scale_x_continuous(
    name = "ERCC estimated cell number", 
    breaks = 0:max(round(ercc$estimatedCellNumber))
  ) +
  scale_y_continuous(
    name = "Detected cell number",
    breaks = 0:max(round(nConnections))
  ) +
  theme_bw()
```

### Fig 9: Detected cell number vs. Total counts
```{r, fig.align="center", fig.width=10, fig.height=8}
tibble(
  sample = names(nConnections),
  detectedConnections = nConnections
) %>%
  inner_join(tibble(
    sample = colnames(getData(cObjMul, "counts")),
    total.counts = colSums(getData(cObjMul, "counts"))
  ), by = "sample") %>%
  ggplot() +
  geom_boxplot(aes(detectedConnections, total.counts, group = detectedConnections)) +
  scale_x_continuous(
    name = "Detected cell number", 
    breaks = 0:max(nConnections)
  ) +
  scale_y_continuous(name = "Total counts") +
  theme_bw()
```

### Fig 10: Detected cell number vs. Total ERCC counts
```{r, fig.align="center", fig.width=10, fig.height=8}
tibble(
  sample = names(nConnections),
  detectedConnections = nConnections
) %>%
  inner_join(tibble(
    sample = colnames(getData(cObjMul, "counts")),
    total.ercc = colSums(getData(cObjMul, "counts.ercc"))
  ), by = "sample") %>%
  ggplot() +
  geom_boxplot(aes(detectedConnections, total.ercc, group = detectedConnections)) +
  scale_x_continuous(
    name = "Detected cell number", 
    breaks = 0:max(nConnections)
  ) +
  scale_y_continuous(name = "Total ERCC counts") +
  theme_bw()
```

### Fig 11: Relative frequency of singlets vs. deconvoluted multiplets

```{r, fig.align="center", fig.width=10, fig.height=8}
singlets <- c(table(getData(cObjSng, "classification")))
singlets <- singlets / sum(singlets)
deconv <- colSums(adjustFractions(cObjSng, cObjMul, sObj))
deconv <- deconv[match(names(singlets), names(deconv))]
deconv <- deconv / sum(deconv)
if(!identical(names(singlets), names(deconv))) stop("name mismatch")

p <- tibble(
  class = names(singlets),
  singlet.freq = singlets,
  multiplet.freq = deconv
) %>%
  ggplot() +
  geom_point(aes(singlet.freq, multiplet.freq, colour = class), size = 3) +
  scale_colour_manual(values = palette('c')[order(names(palette('c')))]) +
  xlim(min(c(deconv, singlets)), max(c(deconv, singlets))) +
  ylim(min(c(deconv, singlets)), max(c(deconv, singlets))) +
  geom_abline(slope = 1, intercept = 0, lty = 3, colour = "grey") +
  labs(x = "Singlet relative frequency", y = "Multiplet relative frequency") +
  guides(colour = guide_legend(title = "Cell Type")) +
  theme_bw()

p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.sngMulRelFreq.pdf',
  device = cairo_pdf,
  height = 180,
  width = 180,
  units = "mm"
)
```

### Fig 12: All connections

```{r, fig.align="center", fig.width=10, fig.height=8}
plotSwarmCircos(
  sObj, cObjSng, cObjMul, classOrder = classOrder.MGA('c'), 
  classColour = palette('c')[classOrder.MGA('c')], h.ratio = 0.85
)
```

### Fig 13: Filtered

Only detected duplicates, triplicates, and quadruplicates.  
ERCC estimated cell number set to max 4.  
Weight cutoff = 10.
Alpha = 1e-3.

```{r, fig.align="center", fig.width=14, fig.height=12}
# adj <- adjustFractions(cObjSng, cObjMul, sObj, binary = TRUE)
# samples <- rownames(adj)
# rs <- rowSums(adj)
# keep <- rs == 2 | rs == 3 | rs == 4

plotSwarmCircos(
  sObj, cObjSng, cObjMul, weightCut = 10, 
  classOrder = classOrder.MGA('c'), theoretical.max = 4, classColour = palette('c')[classOrder.MGA('c')], 
  h.ratio = 0.85, alpha = 1e-3
)

pdf('../figures/MGA.enge20.circos.pdf', width = 9.5, height = 9.5, onefile=FALSE)
plotSwarmCircos(
  sObj, cObjSng, cObjMul, weightCut = 10, 
  classOrder = classOrder.MGA('c'), theoretical.max = 4, classColour = palette('c')[classOrder.MGA('c')], 
  h.ratio = 0.85, alpha = 1e-3
)
dev.off()
```

Calculate the probability of observing Lgr5 expression when Plet1 is or is not 
expressed in Muc2 high expressing multiplets.

```{r, fig.align="center", fig.width=14, fig.height=12}
p <- getData(cObjMul, "counts.cpm") %>% 
  .[c("Plet1", "Lgr5", "Muc2"), ] %>%
  t() %>%
  matrix_to_tibble("sample") %>%
  filter(Muc2 > 3000) %>% #include only Muc2 high
  mutate(
    express.plet1 = if_else(Plet1 > 0, TRUE, FALSE),
    express.lgr5 = if_else(Lgr5 > 0, TRUE, FALSE)
  ) %>%
  group_by(express.plet1, express.lgr5) %>%
  summarize(n = n()) %>%
  ungroup() %>%
  group_by(express.plet1) %>%
  mutate(total = sum(n)) %>%
  ungroup() %>%
  mutate(lgr5.prob = n / total) %>%
  filter(express.lgr5) %>%
  ggplot() + 
  geom_bar(aes(express.plet1, lgr5.prob), stat = "identity", position = position_dodge(width = 1)) +
  labs(x = "Plet1 expressed", y = "Lgr5 probability") +
  ggthemes::theme_few()

p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.Lgr5prob.pdf',
  device = cairo_pdf,
  height = 180,
  width = 90,
  units = "mm"
)
```

```{r, eval = FALSE}
pdata <- adjustFractions(cObjSng, cObjMul, sObj, theoretical.max = 4) %>%
  matrix_to_tibble("sample") %>%
  filter(`Goblet Plet1 1` == 1) %>%
  select(-`Goblet Plet1 1`) %>%
  gather(class, binary, -sample) %>%
  group_by(sample) %>%
  summarize(others = paste(class[binary == 1], collapse = ", ")) %>%
  mutate(others = map(others, ~str_split(.x, ", ")[[1]])) %>%
  unnest() %>%
  filter(others != "") %>%
  group_by(others) %>%
  summarize(prob = n() / nrow(.)) %>% 
  rename(class = others) %>%
  full_join(tibble(class = unique(getData(cObjSng, "classification")))) %>%
  filter(class != "Goblet Plet1 1") %>%
  replace_na(list(prob = 0))

p <- pdata %>% 
  ggplot() +
  geom_bar(aes(class, prob), stat = "identity", position = position_dodge(width = 1)) +
  geom_text(aes(class, prob + 0.01, label = round(prob, digits = 3))) +
  theme_bw() +
  labs(y = "Probability") +
  theme(axis.title.x = element_blank(), axis.text.x = element_text(angle = 90))

p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.PletIntProb.pdf',
  device = cairo_pdf,
  height = 240,
  width = 240,
  units = "mm"
)
```

Plot mean Lgr5 expression in Stem -> colonocyte differentiation trajectory.

```{r, fig.align="center", fig.width=14, fig.height=12}
classes <- c("Stem 1", "Stem 2", "Stem 3", "Transit amplifying", "Progenitor", "Colonocytes")
markers <- c("Lgr5", "Mki67", "Slc26a3")
gene.order <- markers
scale.func <- scale_radius
scale.min <- NA
scale.max <- NA
  
  
getData(cObjSng, "counts.cpm")[markers, ] %>%
  t() %>%
  as.data.frame() %>%
  rownames_to_column("Sample") %>%
  as_tibble() %>% 
  gather(gene, cpm, -Sample) %>%
  inner_join(tibble(
      Sample = colnames(getData(cObjSng, "counts")), 
      Classification = getData(cObjSng, "classification")
  )) %>%
  filter(Classification %in% classes) %>%
  group_by(gene, Classification) %>%
  summarize(mean = mean(cpm), pct = 100 * (length(which(cpm != 0)) / n())) %>%
  mutate(scaled.mean.exp = scale(mean)) %>%
  ungroup() %>%
  mutate(Classification = parse_factor(Classification, levels = classes)) %>%
  mutate(gene = parse_factor(gene, levels = rev(markers))) %>%
  ggplot() +
  geom_point(
      aes(Classification, gene, size = pct, colour = scaled.mean.exp)
  ) +
  scale_colour_gradient(low = "white", high = "darkred") +
  scale.func(range = c(0, 18), limits = c(scale.min, scale.max)) +
  guides(
    size = guide_legend(
      title = "% expressed", title.position = "top", title.hjust = 0.5
    ),
    colour = guide_colorbar(
      title = "Scaled mean expression", title.position = "top", 
      title.hjust = 0.5, barwidth = 10
    )
  ) +
  theme_bw() +
  theme(
    axis.text.x = element_text(angle = 90, hjust = 1), 
    axis.title = element_blank(), 
    legend.position = "top",
    legend.justification = "center"
  )
  
```

Differential expression Plet1 positive vs. Plet1 negative. Show top 25 DE genes.

```{r, fig.align="center", fig.width=10, fig.height=8}
data <- seuratDE %>% 
  rownames_to_column("gene") %>% 
  slice(1:25) %>%
  as.data.frame()

data

#plot selected
markers <- c("Plet1", "Ccl9", "Kcnh3", "Mt3", "Pcsk1", "Cgref1", "Cd44", "Lgals9", "Ang")
p <- plotUnsupervisedMarkers(cObjSng, cObjMul, markers) %>%
  plotData() %>%
  gather(gene, value, -Sample, -(`Sample type`:Colour)) %>%
  mutate(gene = parse_factor(gene, levels = markers)) %>%
  ggplot() +
  geom_point(aes(`dim.red dim 1`, `dim.red dim 2`, colour = value), size = 0.5) +
  scale_colour_viridis_c(option = "E") +
  facet_wrap(~gene, scales = "free") +
  ggthemes::theme_few() +
  theme(legend.position = "top") +
  guides(colour = guide_colourbar(
    title = "log2(CPM + 1) normalized to [0, 1]" ,
    title.position = "top", 
    title.hjust = 0.5,
    barwidth = 15
  ))

p
ggsave(
  plot = p,
  filename = '../figures/MGA.enge20.DE.pdf',
  device = cairo_pdf,
  height = 240,
  width = 240,
  units = "mm"
)

```

```{r}
sessionInfo()
```

