---
title: "sp.scRNAseq sensivity testing"
author: "Jason T. Serviss"
date: "24/10/2016"
output:
  html_document:
    highlight: pygments
    theme: readable
    code_folding: hide
---

```{r setup, echo=FALSE, message=FALSE}
library(sp.scRNAseqTesting)
library(knitr)

##the function below allows dynamic insertion of the function source code 
insert_fun = function(name) {
  read_chunk(lines = capture.output(dump(name, '')), labels = paste(name, 'source', sep = '-'))
}
```

## Introduction
To test the sensivity of the sp.scRNAseq package the "synthetic dataset" was used to run 2 tests. 

1. Multuplet gene expression values were gradually replaced with random values and sp.scRNAseq was run after each replacment.
2. The number of samples (cell types) in each group was gradually reduced and sp.scRNAseq was run after each reduction.

## Synthetic dataset
The synthetic dataset is comprised of 10 cell types, with 100 individual cells per cell type. Each cell has expression values for 2000 genes.

## Test 1

In the first test we gradually replace gene expression values in the multuplets with random noise. For each percentage of random data 10 multuplets consisting of 2, 3, and 4 cells were randomly picked and run. The results indicate that only 20% of the gene expression values are necessary for sp.scRNAseq to determine the correct results.

```{r, fig.align='center', fig.height=8, fig.width=10, eval=TRUE, message=FALSE}
noiseDataTestPlot()
```

## Test 2
In the second test, again using the synthetic dataset, we remove 10 of the samples from each cell type each iteration. sp.scRNAseq is then run for 6 multuplets (2 consisting of 2 cells, 2 consisting of 3 cells, and 2 consisting of 4 cells). The results indicate that performance drops quite quickly and already at 90 samples per cell type sp.scRNAseq is not capable of determining correct results under all conditions.

```{r, fig.align='center', fig.height=8, fig.width=10, eval=TRUE, message=FALSE}
plotGroupSizeTest()
```

## Functions {.tabset}

```{r sourceLoad, echo=FALSE}
insert_fun("noiseDataTest")
.generateNoiseData <- sp.scRNAseqTesting:::.generateNoiseData
insert_fun(".generateNoiseData")
insert_fun("noiseDataTestPlot")
.formatNoiseResults <- sp.scRNAseqTesting:::.formatNoiseResults
insert_fun(".formatNoiseResults")
.addPercentCorrect <- sp.scRNAseqTesting:::.addPercentCorrect
insert_fun(".addPercentCorrect")

insert_fun("groupSizeTest")
insert_fun("plotGroupSizeTest")
.processGroupSizeTest <- sp.scRNAseqTesting:::.processGroupSizeTest
insert_fun(".processGroupSizeTest")
.plotGroupSizeProcessedRes <- sp.scRNAseqTesting:::.plotGroupSizeProcessedRes
insert_fun(".plotGroupSizeProcessedRes")

```


### noiseDataTest
```{r noiseDataTest-source}
```

#### .generateNoiseData
```{r .generateNoiseData-source}
```

### noiseDataTestPlot
```{r noiseDataTestPlot-source}
```

#### .formatNoiseResults
```{r .formatNoiseResults-source}
```

#### .addPercentCorrect
```{r .addPercentCorrect-source}
```

### groupSizeTest
```{r groupSizeTest-source}
```

### plotGroupSizeTest
```{r plotGroupSizeTest-source}
```

#### .processGroupSizeTest
```{r .processGroupSizeTest-source}
```

#### .plotGroupSizeProcessedRes
```{r .plotGroupSizeProcessedRes-source}
```


#
<br></br><br></br><br></br><br></br>


